### ISSUE TITLE
Exception if Client vs. Server evaluation part is used in an OrderBy method

### ISSUE THREAD

**Issue by JonPSmith** [aspnet/EntityFramework#7170]
### Steps to reproduce
I am writing a book on EF Core and I was trying to show how NOT to use Client vs. Server evaluation, i.e. you should not use a client-side evaluated portion in a server-side part of the query. 

Instead of it telling me I can't do (or being slow) I got an exception which talked about duplicate keys.

Obviously this is a VERY low priority issue as people shouldn't be doing this, but I though you might like to know about it.

The Test code is:

```
[Fact]
public void TestClientServerBadPerformanceOk()
{
    //SETUP
    var options =
        this.EnsureUniqueDatabaseWithData();

    using (var db = new EfCoreContext(options))
    {
        var logIt = new LogDbContext(db);

        //ATTEMPT
        var result = db.Books
            .Select(p => new
                {
                    p.Title,
                    p.Price,
                    AuthorsString = string.Join(", ", 
                        p.AuthorsLink 
                            .OrderBy(q => q.Order) 
                            .Select(q => q.Author.Name) 
                    )
                }
            ).OrderBy(b => b.AuthorsString).First();

        //VERIFY
        result.AuthorsString.ShouldNotBeNull();
        foreach (var log in logIt.Logs)
        {
            _output.WriteLine(log);
        }
        //to get the logs you need to fail see https://github.com/aspnet/Tooling/issues/541
        Assert.True(false, "failed the test so that the logs show");
    }
}
```

Where the Book class has a many-to-many linking table to an Author class.
The exception is caused by the `.OrderBy(b => b.AuthorsString)` command (without that it the test works fine).

### The issue
I ran the code listed above expecting an exception saying I could not do that. However I got the exception method.

If you are seeing an exception, include the full exceptions details (message and stack trace).

```
An item with the same key has already been added. Key: join Author q.Author in value(Microsoft.EntityFrameworkCore.Query.Internal.EntityQueryable`1[DataLayer.EfClasses.Author]) on Property([q], "AuthorId") equals Property([q.Author], "AuthorId")
   at System.ThrowHelper.ThrowAddingDuplicateWithKeyArgumentException(Object key)
   at System.Collections.Generic.Dictionary`2.Insert(TKey key, TValue value, Boolean add)
   at Microsoft.EntityFrameworkCore.Query.RelationalQueryModelVisitor.SnapshotQuerySourceMapping(QueryModel queryModel)
   at Microsoft.EntityFrameworkCore.Query.RelationalQueryModelVisitor.OptimizeJoinClause(JoinClause joinClause, QueryModel queryModel, Int32 index, Action baseVisitAction, MethodInfo operatorToFlatten, Boolean groupJoin)
   at Microsoft.EntityFrameworkCore.Query.RelationalQueryModelVisitor.VisitJoinClause(JoinClause joinClause, QueryModel queryModel, Int32 index)
   at Remotion.Linq.QueryModelVisitorBase.VisitBodyClauses(ObservableCollection`1 bodyClauses, QueryModel queryModel)
   at Remotion.Linq.QueryModelVisitorBase.VisitQueryModel(QueryModel queryModel)
   at Microsoft.EntityFrameworkCore.Query.EntityQueryModelVisitor.VisitQueryModel(QueryModel queryModel)
   at Microsoft.EntityFrameworkCore.Query.RelationalQueryModelVisitor.VisitQueryModel(QueryModel queryModel)
   at Microsoft.EntityFrameworkCore.Query.Internal.SqlServerQueryModelVisitor.VisitQueryModel(QueryModel queryModel)
   at Microsoft.EntityFrameworkCore.Query.ExpressionVisitors.RelationalEntityQueryableExpressionVisitor.VisitSubQuery(SubQueryExpression expression)
   at Microsoft.EntityFrameworkCore.Query.ExpressionVisitors.ExpressionVisitorBase.Visit(Expression node)
   at System.Dynamic.Utils.ExpressionVisitorUtils.VisitArguments(ExpressionVisitor visitor, IArgumentProvider nodes)
   at System.Linq.Expressions.ExpressionVisitor.VisitMethodCall(MethodCallExpression node)
   at System.Linq.Expressions.MethodCallExpression.Accept(ExpressionVisitor visitor)
   at Microsoft.EntityFrameworkCore.Query.ExpressionVisitors.ExpressionVisitorBase.Visit(Expression node)
   at Microsoft.EntityFrameworkCore.Query.EntityQueryModelVisitor.ReplaceClauseReferences(Expression expression, IQuerySource querySource, Boolean inProjection)
   at Microsoft.EntityFrameworkCore.Query.EntityQueryModelVisitor.VisitOrdering(Ordering ordering, QueryModel queryModel, OrderByClause orderByClause, Int32 index)
   at Remotion.Linq.QueryModelVisitorBase.VisitOrderings(ObservableCollection`1 orderings, QueryModel queryModel, OrderByClause orderByClause)
   at Remotion.Linq.QueryModelVisitorBase.VisitBodyClauses(ObservableCollection`1 bodyClauses, QueryModel queryModel)
   at Remotion.Linq.QueryModelVisitorBase.VisitQueryModel(QueryModel queryModel)
   at Microsoft.EntityFrameworkCore.Query.EntityQueryModelVisitor.VisitQueryModel(QueryModel queryModel)
   at Microsoft.EntityFrameworkCore.Query.RelationalQueryModelVisitor.VisitQueryModel(QueryModel queryModel)
   at Microsoft.EntityFrameworkCore.Query.Internal.SqlServerQueryModelVisitor.VisitQueryModel(QueryModel queryModel)
   at Microsoft.EntityFrameworkCore.Query.EntityQueryModelVisitor.CreateQueryExecutor[TResult](QueryModel queryModel)
   at Microsoft.EntityFrameworkCore.Query.Internal.QueryCompiler.CompileQueryCore[TResult](Expression query, INodeTypeProvider nodeTypeProvider, IDatabase database, ILogger logger, Type contextType)
   at Microsoft.EntityFrameworkCore.Query.Internal.QueryCompiler.<>c__DisplayClass19_0`1.<CompileQuery>b__0()
   at Microsoft.EntityFrameworkCore.Query.Internal.CompiledQueryCache.GetOrAddQueryCore[TFunc](Object cacheKey, Func`1 compiler)
   at Microsoft.EntityFrameworkCore.Query.Internal.QueryCompiler.Execute[TResult](Expression query)
   at System.Linq.Queryable.First[TSource](IQueryable`1 source)
   at test.UnitTests.TestEfCoreContext.TestClientServerBadPerformanceOk() in C:\Users\Jon\Documents\Visual Studio 2015\Projects\EfCoreInAction\test\UnitTests\TestEfCoreContext.cs:line 279
```

### Further technical details

EF Core version: 1.1.0
Operating system: Windows 10
Visual Studio version: VS2015 update 3

Other details about my project setup:

**Comment by JonPSmith** [comment]
Note: I tried another example and that failed differently

Same Unit Test framework but ATTEMPT Part was
```
//ATTEMPT
var result = db.Books
    .Select(p => new
        {
            p.Title,
            p.Price,
            CalcVal = p.Title.Length
        }
    ).OrderBy(b => b.CalcVal).First();
```

Exception this time was
```
Unable to cast object of type 'System.Int64' to type 'System.Int32'.
   at lambda_method(Closure , ValueBuffer )
   at System.Linq.Enumerable.SelectEnumerableIterator`2.MoveNext()
   at Microsoft.EntityFrameworkCore.Query.Internal.LinqOperatorProvider.ExceptionInterceptor`1.EnumeratorExceptionInterceptor.MoveNext()
   at System.Collections.Generic.List`1..ctor(IEnumerable`1 collection)
   at System.Linq.Enumerable.ToList[TSource](IEnumerable`1 source)
   at test.UnitTests.TestEfCoreContext.TestClientServerBadPerformanceOk() in C:\Users\Jon\Documents\Visual Studio 2015\Projects\EfCoreInAction\test\UnitTests\TestEfCoreContext.cs:line 281
```

Same setup

**Comment by maumar** [comment]
The first scenario works in the current bits. We produce the following query plan (with join being correctly executed on the client):

```
(QueryContext queryContext) => IEnumerable<<>f__AnonymousType0<string, decimal, string>> _InterceptExceptions(
    source: IEnumerable<<>f__AnonymousType0<string, decimal, string>> _ToSequence(<>f__AnonymousType0<string, decimal, string> FirstOrDefault(IEnumerable<<>f__AnonymousType0<string, decimal, string>> _Select(
                source: IOrderedEnumerable<ValueBuffer> _OrderBy(
                    source: IEnumerable<ValueBuffer> _ShapedQuery(
                        queryContext: queryContext, 
                        shaperCommandContext: SelectExpression: 
                            SELECT [p].[Id], [p].[Title], [p].[Price]
                            FROM [Books] AS [p], 
                        shaper: ValueBufferShaper), 
                    expression: (ValueBuffer p) => string Join(
                        separator: ", ", 
                        values: IEnumerable<string> _Select(
                            source: IEnumerable<TransparentIdentifier<TransparentIdentifier<ValueBuffer, ValueBuffer>, ValueBuffer>> _InjectParameters(
                                queryContext: queryContext, 
                                source: IEnumerable<TransparentIdentifier<TransparentIdentifier<ValueBuffer, ValueBuffer>, ValueBuffer>> _ShapedQuery(
                                    queryContext: queryContext, 
                                    shaperCommandContext: SelectExpression: 
                                        SELECT [q].[AuthorId], [q.Author].[Name]
                                        FROM [BookAuthors] AS [q]
                                        INNER JOIN [Authors] AS [q.Author] ON [q].[AuthorId] = [q.Author].[Id]
                                        WHERE @_outer_Id = [q].[BookId]
                                        ORDER BY [q].[Order], 
                                    shaper: TypedCompositeShaper<TypedCompositeShaper<ValueBufferShaper, ValueBuffer, ValueBufferShaper, ValueBuffer, TransparentIdentifier<ValueBuffer, ValueBuffer>>, TransparentIdentifier<ValueBuffer, ValueBuffer>, ValueBufferShaper, ValueBuffer, TransparentIdentifier<TransparentIdentifier<ValueBuffer, ValueBuffer>, ValueBuffer>>), 
                                parameterNames: new string[]{ "_outer_Id" }, 
                                parameterValues: new object[]{ (object) Nullable<int> TryReadValue(p, 0, Book.Id) }), 
                            selector: (TransparentIdentifier<TransparentIdentifier<ValueBuffer, ValueBuffer>, ValueBuffer> t1) => string TryReadValue(t1.Inner, 1, Author.Name))), 
                    orderingDirection: Asc), 
                selector: (ValueBuffer p) => new <>f__AnonymousType0<string, decimal, string>(
                    string TryReadValue(p, 1, Book.Title), 
                    decimal TryReadValue(p, 2, Book.Price), 
                    string Join(
                        separator: ", ", 
                        values: IEnumerable<string> _Select(
                            source: IEnumerable<TransparentIdentifier<TransparentIdentifier<ValueBuffer, ValueBuffer>, ValueBuffer>> _InjectParameters(
                                queryContext: queryContext, 
                                source: IEnumerable<TransparentIdentifier<TransparentIdentifier<ValueBuffer, ValueBuffer>, ValueBuffer>> _ShapedQuery(
                                    queryContext: queryContext, 
                                    shaperCommandContext: SelectExpression: 
                                        SELECT [q0].[AuthorId], [q.Author1].[Name]
                                        FROM [BookAuthors] AS [q0]
                                        INNER JOIN [Authors] AS [q.Author1] ON [q0].[AuthorId] = [q.Author1].[Id]
                                        WHERE @_outer_Id1 = [q0].[BookId]
                                        ORDER BY [q0].[Order], 
                                    shaper: TypedCompositeShaper<TypedCompositeShaper<ValueBufferShaper, ValueBuffer, ValueBufferShaper, ValueBuffer, TransparentIdentifier<ValueBuffer, ValueBuffer>>, TransparentIdentifier<ValueBuffer, ValueBuffer>, ValueBufferShaper, ValueBuffer, TransparentIdentifier<TransparentIdentifier<ValueBuffer, ValueBuffer>, ValueBuffer>>), 
                                parameterNames: new string[]{ "_outer_Id1" }, 
                                parameterValues: new object[]{ (object) Nullable<int> TryReadValue(p, 0, Book.Id) }), 
                            selector: (TransparentIdentifier<TransparentIdentifier<ValueBuffer, ValueBuffer>, ValueBuffer> t1) => string TryReadValue(t1.Inner, 1, Author.Name)))
                )))), 
    contextType: Repro7170.MyContext, 
    logger: InterceptingLogger<Query>, 
    queryContext: queryContext)
```

**Comment by maumar** [comment]
filed https://github.com/aspnet/EntityFramework/issues/8176 for the second issue

**Comment by JonPSmith** [comment]
Hi @maumar,

Yes, I checked. This was fixed in 1.1.1
